File: references/samples/java/src/main/java/com/codevasp/utils/CryptoUtils.java
package com.codevasp.utils;

import com.goterl.lazysodium.LazySodiumJava;
import com.goterl.lazysodium.SodiumJava;
import com.goterl.lazysodium.exceptions.SodiumException;
import com.goterl.lazysodium.interfaces.Box;
import com.goterl.lazysodium.interfaces.MessageEncoder;
import com.goterl.lazysodium.interfaces.Sign;
import com.goterl.lazysodium.utils.KeyPair;
import lombok.Getter;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.time.ZoneId;
import java.util.*;

public class CryptoUtils {

    private static final LazySodiumJava sodium;

    static {
        sodium = new LazySodiumJava(new SodiumJava(), new UrlSafeBase64MessageEncoder());
        sodium.sodiumInit();
    }

    public static class UrlSafeBase64MessageEncoder implements MessageEncoder {

        @Override
        public String encode(byte[] cipher) {
            return Base64.getEncoder().encodeToString(cipher);
        }

        @Override
        public byte[] decode(String cipherText) {
            return Base64.getDecoder().decode(cipherText);
        }
    }

    public static String generateSignKey() {
        try {
            KeyPair newKey = sodium.cryptoSignKeypair();
            byte[] key64 = newKey.getSecretKey().getAsBytes();
            byte[] key32 = new byte[32];
            System.arraycopy(key64, 0, key32, 0, 32);
            return Base64.getEncoder().encodeToString(key32);
        } catch (Exception e) {
            return null;
        }
    }

    public static byte[] decrypt(byte[] data, byte[] sharedKey) throws SodiumException {
        byte[] nonce = Arrays.copyOfRange(data, 0, Box.NONCEBYTES);
        byte[] encryptedMsg = Arrays.copyOfRange(data, Box.NONCEBYTES, data.length);
        byte[] decrypted = new byte[encryptedMsg.length - Box.MACBYTES];
        if (!sodium.cryptoBoxOpenEasyAfterNm(
                decrypted, encryptedMsg, encryptedMsg.length, nonce, sharedKey)) {
            throw new SodiumException("Could not decrypt data");
        }
        return decrypted;
    }

    public static String decrypt(String data, String selfPrivateKey, String remotePublicKey)
            throws SodiumException {
        byte[] dataBytes = Base64.getDecoder().decode(data);
        byte[] sharedKey = getSharedKey(remotePublicKey, selfPrivateKey);
        return new String(decrypt(dataBytes, sharedKey));
    }

    private static byte[] getSharedKey(byte[] publicKey, byte[] privateKey) {
        byte[] sharedKey = new byte[Box.BEFORENMBYTES];
        byte[] curPublicKey = new byte[Sign.CURVE25519_PUBLICKEYBYTES];
        byte[] curPrivateKey = new byte[Sign.CURVE25519_SECRETKEYBYTES];
        sodium.convertPublicKeyEd25519ToCurve25519(curPublicKey, publicKey);
        sodium.convertSecretKeyEd25519ToCurve25519(curPrivateKey, privateKey);
        sodium.cryptoBoxBeforeNm(sharedKey, curPublicKey, curPrivateKey);
        return sharedKey;
    }

    private static byte[] getSharedKey(String publicKey, String privateKey) {
        byte[] pukBytes = Base64.getDecoder().decode(publicKey);
        byte[] prkBytes = Base64.getDecoder().decode(privateKey);
        return getSharedKey(pukBytes, prkBytes);
    }

    public static byte[] encrypt(byte[] data, byte[] sharedKey) throws SodiumException, IOException {
        byte[] nonce = sodium.randomBytesBuf(Box.NONCEBYTES);
        byte[] encrypted = new byte[data.length + Box.MACBYTES];
        if (!sodium.cryptoBoxEasyAfterNm(encrypted, data, data.length, nonce, sharedKey)) {
            throw new SodiumException("Could not encrypt data");
        }
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        out.write(nonce);
        out.write(encrypted);
        return out.toByteArray();
    }

    public static String encrypt(String data, String selfPrivateKey, String remotePublicKey)
            throws SodiumException, IOException {
        byte[] dataBytes = data.getBytes();
        byte[] sharedKey = getSharedKey(remotePublicKey, selfPrivateKey);
        return Base64.getEncoder().encodeToString(encrypt(dataBytes, sharedKey));
    }

    @Getter
    public static class SignatureContainer {
        // Getters
        private final Integer iNonce;
        private final String signatureBase64;
        private final String strDateTime;

        public SignatureContainer(Integer iNonce, String signatureBase64, String strDateTime) {
            this.iNonce = iNonce;
            this.signatureBase64 = signatureBase64;
            this.strDateTime = strDateTime;
        }
    }

    public static SignatureContainer generateSignature(String body, String ownPrivateKey) throws IOException, SodiumException {
        // Generating Signature
        int iNonce = (int) (Math.random() * 10000);

        Calendar calendar = Calendar.getInstance();
        Date date = calendar.getTime();
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ");
        sdf.setTimeZone(TimeZone.getTimeZone(ZoneId.of("UTC")));
        String strDateTime = sdf.format(date);

        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        buffer.write(strDateTime.getBytes());
        buffer.write(body.getBytes());
        buffer.write(toBytes(iNonce));

        byte[] signatureBytes = new byte[Sign.BYTES];
        byte[] seed = Base64.getDecoder().decode(ownPrivateKey);
        KeyPair signKeyPair = sodium.cryptoSignSeedKeypair(seed);
        sodium.cryptoSignDetached(signatureBytes, buffer.toByteArray(), buffer.toByteArray().length,
                signKeyPair.getSecretKey().getAsBytes());

        String signatureBase64 = Base64.getEncoder().encodeToString(signatureBytes);

        return new SignatureContainer(iNonce, signatureBase64, strDateTime);
    }

    public static String getPublicKey(String privateKey) throws SodiumException {
        byte[] seed = Base64.getDecoder().decode(privateKey);
        KeyPair signKeyPair = sodium.cryptoSignSeedKeypair(seed);
        return Base64.getEncoder().encodeToString(signKeyPair.getPublicKey().getAsBytes());
    }

    private static byte[] toBytes(int i) {
        byte[] result = new byte[4];

        result[0] = (byte) (i >> 24);
        result[1] = (byte) (i >> 16);
        result[2] = (byte) (i >> 8);
        result[3] = (byte) (i /*>> 0*/);

        return result;
    }
}

File: references/samples/go/utils/crypto_utils.go
package crypto

import (
	"bytes"
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"golang.org/x/crypto/ed25519"
	math "math/rand"
	"time"

	"golang.org/x/crypto/nacl/box"
)

type Curve25519 struct {
	PrivateKey *[32]byte
	PublicKey  *[32]byte
}

func PaddingShift(bb []byte, size int) []byte {
	l := len(bb)
	if l == size {
		return bb
	}
	if l > size {
		return bb[l-size:]
	}
	tmp := make([]byte, size)
	copy(tmp[:size-l], bb)
	return tmp
}

func GetPublicKey(base64OwnPrivateKey string) (string, error) {
	// Decode the base64 private key
	privateKeyBytes, err := base64.StdEncoding.DecodeString(base64OwnPrivateKey)
	if err != nil {
		return "", fmt.Errorf("failed to decode private key: %v", err)
	}

	// Ensure the private key is 32 bytes (seed)
	if len(privateKeyBytes) != ed25519.SeedSize {
		return "", fmt.Errorf("invalid private key length: got %d, want %d", len(privateKeyBytes), ed25519.SeedSize)
	}

	// Generate the full private key from the seed
	privateKey := ed25519.NewKeyFromSeed(privateKeyBytes)

	// Get the public key
	publicKey := privateKey.Public().(ed25519.PublicKey)

	// Encode the public key to base64
	base64PublicKey := base64.StdEncoding.EncodeToString(publicKey)

	return base64PublicKey, nil
}

func Sodium(base64PrivateKey string, base64PublicKey string) (*Curve25519, error) {
	privBytes, _ := base64.StdEncoding.DecodeString(base64PrivateKey)
	pubBytes, _ := base64.StdEncoding.DecodeString(base64PublicKey)

	var privateKey [32]byte
	var publicKey [32]byte

	copy(privateKey[:], privBytes)
	copy(publicKey[:], pubBytes)

	return &Curve25519{PrivateKey: &privateKey, PublicKey: &publicKey}, nil
}

func (cc *Curve25519) Encrypt(msg string) ([]byte, error) {
	var nonce [24]byte
	_, err := rand.Read(nonce[:])
	if err != nil {
		return nil, err
	}

	encrypted := box.Seal(nonce[:], []byte(msg), &nonce, cc.PublicKey, cc.PrivateKey)
	return encrypted, nil
}

func (cc *Curve25519) Decrypt(encrypted []byte) ([]byte, error) {
	var Nonce [24]byte
	copy(Nonce[:], encrypted[:24])

	decrypted, success := box.Open(nil, encrypted[24:], &Nonce, cc.PublicKey, cc.PrivateKey)
	if !success {
		return nil, fmt.Errorf("Failed to decrypt message")
	}

	return decrypted, nil
}

type SignatureContainer struct {
	Nonce     int
	Signature string
	DateTime  string
}

func GenerateSignature(body string, ownPrivateKey string) (SignatureContainer, error) {

	r := math.New(math.NewSource(time.Now().UnixNano()))
	iNonce := r.Intn(10000)

	now := time.Now().UTC()
	strDateTime := now.Format("2006-01-02T15:04:05-0700")

	buffer := new(bytes.Buffer)
	buffer.Write([]byte(strDateTime))
	buffer.Write([]byte(body))
	buffer.Write(ToBytes(iNonce))

	// Decode the base64 private key
	privateKeyBytes, err := base64.StdEncoding.DecodeString(ownPrivateKey)
	if err != nil {
		panic(err)
	}

	var privateKey ed25519.PrivateKey
	var _ ed25519.PublicKey

	switch len(privateKeyBytes) {
	case 32:
		// If it's 32 bytes, it might be a seed for Ed25519
		privateKey = ed25519.NewKeyFromSeed(privateKeyBytes)
		_ = privateKey.Public().(ed25519.PublicKey)
	case 64:
		// If it's 64 bytes, it's likely a full Ed25519 private key
		privateKey = privateKeyBytes
		_ = privateKey.Public().(ed25519.PublicKey)
	default:
		panic(fmt.Sprintf("Unexpected private key length: %d bytes", len(privateKeyBytes)))
	}

	// Sign the message
	signature := ed25519.Sign(privateKey, buffer.Bytes())
	signatureBase64 := base64.StdEncoding.EncodeToString(signature)

	return SignatureContainer{
		Nonce:     iNonce,
		Signature: signatureBase64,
		DateTime:  strDateTime,
	}, nil
}

func ToBytes(i int) []byte {
	result := make([]byte, 4)

	result[0] = byte(i >> 24)
	result[1] = byte(i >> 16)
	result[2] = byte(i >> 8)
	result[3] = byte(i)

	return result
}
